### PR TITLE
New OpenXR ActionSets and InteractionProfiles assets

### PR DESCRIPTION
**Pull request by galibzon**
## What does this PR do?

New OpenXR ActionSets and InteractionProfiles assets per this RFC:
https://github.com/o3de/sig-graphics-audio/issues/156
This commit is the first of several more that will be committed in different batches for the sake of making it easier for code reviewers.

This commit has no runtime implications, other that registering the `OpenXRInteractionProfilesAsset` and the `OpenXRActionSetsAsset`, along with the asset builder named `OpenXRAssetsBuilder`, which is in charge
 of validating and building both asset types.

`OpenXRInteractionProfilesAsset` will have the `*.xrprofiles` file extension. An example asset is provided as the following path: `Gems/OpenXRVk/Assets/OpenXRVk/system.xrprofiles`

`OpenXRActionSetsAsset` will have the `*.rxactions` file extension. An example asset is provided as the following path: `Gems/OpenXRVk/Assets/OpenXRVk/default.xractions`
This asset has a job dependency on `Gems/OpenXRVk/Assets/OpenXRVk/system.xrprofiles`.

Both asset types are also editable with the `Asset Editor`.
image

The Assets mentioned above will be compiled and placed in the asset cache but they are unused at the moment for the sake of an easier code review.

## How was this PR tested?

Validated with OpenXRTest and proprietary project using Quest Pro and Quest 3 devices.

### CODE REVIEW

File: Gems/OpenXRVk/Code/Source/Builders/OpenXRVkAssetsBuilderSystemComponent.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        dependent.push_back(AZ_CRC("AssetCatalogService"));
+    }
+    
+    void OpenXRAssetsBuilderSystemComponent::Init()
+    {
+    }
+    
+    void OpenXRAssetsBuilderSystemComponent::Activate()
+    {    
+        // Register Shader Asset Builder
```
Comment (byrcolin) [inline Gems/OpenXRVk/Code/Source/Builders/OpenXRVkAssetsBuilderSystemComponent.cpp:59]: Comment seems incorrect

File: Gems/OpenXRVk/Code/Source/Builders/OpenXRVkAssetsBuilderSystemComponent.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    
+        static void Reflect(AZ::ReflectContext* context);
+    
+        static void GetProvidedServices(AZ::ComponentDescriptor::DependencyArrayType& provided);
+        static void GetIncompatibleServices(AZ::ComponentDescriptor::DependencyArrayType& incompatible);
+        static void GetRequiredServices(AZ::ComponentDescriptor::DependencyArrayType& required);
+        static void GetDependentServices(AZ::ComponentDescriptor::DependencyArrayType& dependent);
+    
+    protected:
+        ////////////////////////////////////////////////////////////////////////
```
Comment (byrcolin) [inline Gems/OpenXRVk/Code/Source/Builders/OpenXRVkAssetsBuilderSystemComponent.h:31]: nit: empty comment section seems incorrect

File: Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp [hunk 3]
```diff
@@ -0,0 +1,608 @@
...
+
+    static AZ::Outcome<void, AZStd::string> ValidateUserPathDescriptor(const OpenXRVk::OpenXRInteractionUserPathDescriptor& userPathDescriptor,
+        AZStd::unordered_set<AZStd::string>& uniqueUserPathNames, AZStd::unordered_set<AZStd::string>& uniqueUserPathPaths,
+        AZStd::unordered_set<AZStd::string>& uniqueComponentPathNames, AZStd::unordered_set<AZStd::string>& uniqueComponentPathPaths)
+    {
+        {
+            if (uniqueUserPathNames.contains(userPathDescriptor.m_name))
+            {
+                return AZ::Failure(
+                    AZStd::string::format("An User Path with name [%s] already exists.",
```
Comment (byrcolin) [inline Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp:173]: I would remove "An"

File: Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp [hunk 4]
```diff
@@ -0,0 +1,608 @@
...
+                return AZ::Failure(
+                    AZStd::string::format("User Path Name [%s] is invalid. Reason:\n%s", userPathDescriptor.m_name.c_str(), outcome.GetError().c_str())
+                );
+            }
+        }
+        {
+            if (uniqueUserPathPaths.contains(userPathDescriptor.m_path))
+            {
+                return AZ::Failure(
+                    AZStd::string::format("An User Path with path [%s] already exists.",
```
Comment (byrcolin) [inline Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp:190]: I would remove "An"

File: Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+
+    static AZ::Outcome<void, AZStd::string> ValidateInteractionProfileDescriptor(
+        const OpenXRVk::OpenXRInteractionProfileDescriptor& interactionProfileDescriptor,
+        AZStd::unordered_set<AZStd::string>& uniqueNames, AZStd::unordered_set<AZStd::string>& uniquePaths)
+    {
+        {
+            if (uniqueNames.contains(interactionProfileDescriptor.m_name))
+            {
+                return AZ::Failure(
+                    AZStd::string::format("An Interaction Profile with name [%s] already exists.",
```
Comment (byrcolin) [inline Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp:227]: I would remove "An"

File: Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp [hunk 6]
```diff
@@ -0,0 +1,608 @@
...
+                    AZStd::string::format("Interaction Profile Unique Name [%s] is invalid. Reason:\n%s", interactionProfileDescriptor.m_name.c_str(), outcome.GetError().c_str())
+                );
+            }
+        }
+
+        {
+            if (uniquePaths.contains(interactionProfileDescriptor.m_path))
+            {
+                return AZ::Failure(
+                    AZStd::string::format("An Interaction Profile with path [%s] already exists.",
```
Comment (byrcolin) [inline Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp:245]: I would remove "An"

File: Gems/OpenXRVk/Code/Source/Builders/OpenXRVkBuilderModule.cpp [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        : public AZ::Module
+    {
+    public:
+        AZ_RTTI(OpenXRBuilderModule, "{43370465-DBF1-44BB-968D-97C0B42F5EA0}", AZ::Module);
+        AZ_CLASS_ALLOCATOR(OpenXRBuilderModule, AZ::SystemAllocator);
+    
+        OpenXRBuilderModule()
+        {
+            // Push results of [MyComponent]::CreateDescriptor() into m_descriptors here.
+            m_descriptors.insert(m_descriptors.end(), {
```
Comment (byrcolin) [inline Gems/OpenXRVk/Code/Source/Builders/OpenXRVkBuilderModule.cpp:26]: nit: Inconstant comment styles

File: Gems/OpenXRVk/Code/Source/OpenXRVkActionSetsAsset.cpp [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+                ->Field("Name", &OpenXRActionDescriptor::m_name)
+                ->Field("LocalizedName", &OpenXRActionDescriptor::m_localizedName)
+                ->Field("Comment", &OpenXRActionDescriptor::m_comment)
+                ->Field("ActionPathDescriptors", &OpenXRActionDescriptor::m_actionPathDescriptors)
+                ;
+
+            AZ::EditContext* edit = serialize->GetEditContext();
+            if (edit)
+            {
+                edit->Class<OpenXRActionDescriptor>("OpenXRActionDescriptor", "An action bound to one or more OpenXR Action Paths.")
```
Comment (byrcolin) [inline Gems/OpenXRVk/Code/Source/OpenXRVkActionSetsAsset.cpp:241]: I would remove "An"

File: Gems/OpenXRVk/Code/Source/OpenXRVkActionSetsAsset.cpp [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+                ->Field("LocalizedName", &OpenXRActionSetDescriptor::m_localizedName)
+                ->Field("Priority", &OpenXRActionSetDescriptor::m_priority)
+                ->Field("Comment", &OpenXRActionSetDescriptor::m_comment)
+                ->Field("ActionDescriptors", &OpenXRActionSetDescriptor::m_actionDescriptors)
+                ;
+
+            AZ::EditContext* edit = serialize->GetEditContext();
+            if (edit)
+            {
+                edit->Class<OpenXRActionSetDescriptor>("OpenXRActionSetDescriptor", "A group of OpenXR Actions that can be selectively activated/deactivated at runtime.")
```
Comment (byrcolin) [inline Gems/OpenXRVk/Code/Source/OpenXRVkActionSetsAsset.cpp:301]: I would remove "A"

File: Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp [hunk 10]
```diff
@@ -0,0 +1,608 @@
...
+        return AZ::Success();
+    }
+
+
+    AZ::Outcome<void, AZStd::string> ValidateInteractionProfilesAsset(
+        const OpenXRVk::OpenXRInteractionProfilesAsset& interactionProfilesAsset)
+    {
+        if (interactionProfilesAsset.m_interactionProfileDescriptors.empty())
+        {
+            return AZ::Failure("An InteractionProfiles asset requires at least one Interaction Profile");
```
Comment (byrcolin) [inline Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp:296]: nit: I would remove "An"

Not going to mark any more of these.

**Review by byrcolin — APPROVED** [review]
only nits

**Comment by galibzon** [comment]
> only nits

Thanks a lot for the feedback.

File: Gems/OpenXRVk/Code/Source/Builders/OpenXRVkAssetsBuilderSystemComponent.cpp (hunk 1 above)
Comment (galibzon) [reply Gems/OpenXRVk/Code/Source/Builders/OpenXRVkAssetsBuilderSystemComponent.cpp:59]: Fixed.

File: Gems/OpenXRVk/Code/Source/Builders/OpenXRVkAssetsBuilderSystemComponent.h (hunk 2 above)
Comment (galibzon) [reply Gems/OpenXRVk/Code/Source/Builders/OpenXRVkAssetsBuilderSystemComponent.h:31]: Fixed.

File: Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp (hunk 4 above)
Comment (galibzon) [reply Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp:190]: Fixed.

File: Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp (hunk 5 above)
Comment (galibzon) [reply Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp:227]: Fixed.

File: Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp (hunk 6 above)
Comment (galibzon) [reply Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp:245]: Fixed.

File: Gems/OpenXRVk/Code/Source/Builders/OpenXRVkBuilderModule.cpp (hunk 7 above)
Comment (galibzon) [reply Gems/OpenXRVk/Code/Source/Builders/OpenXRVkBuilderModule.cpp:26]: File was removed due to CMake target refactoring.

File: Gems/OpenXRVk/Code/Source/OpenXRVkActionSetsAsset.cpp (hunk 8 above)
Comment (galibzon) [reply Gems/OpenXRVk/Code/Source/OpenXRVkActionSetsAsset.cpp:241]: Fixed.

File: Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp (hunk 10 above)
Comment (galibzon) [reply Gems/OpenXRVk/Code/Source/OpenXRVkAssetsValidator.cpp:296]: Fixed.

File: Gems/OpenXRVk/Code/Source/Builders/OpenXRVkAssetsBuilder.cpp [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+           response.m_resultCode = AssetBuilderSDK::ProcessJobResult_Failed;
+           return;
+       }
+
+       AssetBuilderSDK::JobProduct jobProduct;
+       if (!AssetBuilderSDK::OutputObject(interactionProfilesAssetPtr.get(), assetOutputPath,
+           azrtti_typeid<OpenXRVk::OpenXRInteractionProfilesAsset>(),
+           aznumeric_cast<uint32_t>(0), jobProduct))
+       {
+           AZ_Error(LogName, false, "FIXME this message.");
```
Comment (antonmic) [inline Gems/OpenXRVk/Code/Source/Builders/OpenXRVkAssetsBuilder.cpp:136]: Is FIXME a placeholder? Can you improve the error message?

**Review by antonmic — APPROVED** [review]

**Review by moudgils — APPROVED** [review]

File: Gems/OpenXRVk/Code/Source/Builders/OpenXRVkAssetsBuilder.cpp (hunk 11 above)
Comment (galibzon) [reply Gems/OpenXRVk/Code/Source/Builders/OpenXRVkAssetsBuilder.cpp]: Fixed.

**Pull request merged**